560. subarray sum equals k

5ky7 opened the pull request:
問題：https://leetcode.com/problems/subarray-sum-equals-k/

5ky7 committed:
Step 1

5ky7 committed:
Minor change for Step 1

5ky7 committed:
Minor change for Step 1

5ky7 committed:
Revised Step 1

5ky7 committed:
Revised Step 1

5ky7 committed:
Revised Step 1

5ky7 committed:
step 1の修正＋step 2

potrue commented on memo.md:
```diff
@@ -1,3 +1,146 @@
+        vector<int> cumulative_sums(n + 1);
+        cumulative_sums[0] = 0;
+        for (int i = 1; i <= n; ++i) {
+            cumulative_sums[i] = nums[i-1] + cumulative_sums[i-1];
```
numsの各要素についてfor文を回して、cumulative_sums.push_back(cumulative_sums.back() + num)を行ってcumulative_sumsを作ってもよいと思いました。

5ky7 replied:
`vector<int> cumulative_sums` --> `push_back`だと再確保で遅いのかなあと思っていましたが，先に`.reserve()`して`push_back`すれば遅くもならないし可読性もあがりそうです．
ありがとうございます！

potrue commented on memo.md:
```diff
@@ -1,3 +1,146 @@
+              - ただまだ実行速度に700msくらいの差はあって，leetcodeの実行時間が不正確といえどデータ依存による速度の差はちゃんと反映されているようだ．
+      - 大体~10^3msくらいが単純にsubarrayを全て舐めるアルゴリズムの限界？
+        - 計算量だけみるとifの条件文でkと等しいかの比較がsubarraysの数，つまりO(n^2)だけ必要になる．定数倍は上記のような工夫で下げられるが，オーダーをこれ以下に抑えるには根本的に別の考え方が必要．
+    - Hash Mapを使った高速化（[4つ目のコード](#Code4), accepted）．平均時間計算量O(n)，空間計算量O(n)．
```
mapはHashMapではなかった気がします。unordered_mapがHashMapで、mapはまた違うものだと記憶しています。
https://cpprefjp.github.io/reference/map/map.html
https://cpprefjp.github.io/reference/unordered_map/unordered_map.html

5ky7 replied:
ご指摘ありがとうございます！修正しました．
おっしゃる通りで，大事なのはhash mapかどうかより累積和を上手く使うという点でした．

potrue commented on memo.md:
```diff
@@ -1,3 +1,146 @@
+    int subarraySum(vector<int>& nums, int k) {
+        map<int, int> sum_subarray_to_count_subarray;
+        sum_subarray_to_count_subarray[0] = 1;
+        int solution = 0;
```
solutionという変数名で結果を表す変数を定義するのはあまり見ない気がします。countなど文脈固有の変数名ではないものであれば、resultが一般的ではないでしょうか。

potrue commented on memo.md:
```diff
@@ -1,3 +1,146 @@
+class Solution {
+public:
+    int subarraySum(vector<int>& nums, int k) {
+        map<int, int> sum_subarray_to_count_subarray;
```
sum_to_countで十分な気がします。sum(of)とcount(of)が省略されているので、to_count_subarrayが「subarrayをcountするためのsum_subarray」みたいな意味に聞こえてしまって一瞬戸惑いました。

naoto-iwase commented on memo.md:
```diff
@@ -1,3 +1,146 @@
+*以下の文章内ではLeetCodeの計算時間を参考にしている箇所がありますが，実際にはLeetcodeの計測時間は不正確とのことですので，読み飛ばしていただけたらと思います（後にベンチマークでの計測値に置き換える予定です）* 
 # Step 1
+## とりあえずnaiveに実装
+  - 全ての部分列を順に見ていって，和が`k`に等しいかみていく．最悪時間計算量はO(n^2)．空間計算量はO(1)．（以下[1つ目のコード](#Code1)．Time Limit Exceeded）．
```
二重ループの内側で毎回長さ j の reduce を実行しているので時間計算量は O(n^3) でしょうか。

5ky7 replied:
おっしゃる通りです．ご指摘ありがとうございます！

naoto-iwase commented on memo.md:
```diff
@@ -1,3 +1,146 @@
+        int count = 0;
+        for (int i = 0; i < nums.size(); ++i) { // iはスタート位置
+            int sum = 0;
+            for (int j = 0; j < nums.size() - i; ++j) { // jはsubarrayの要素数
+                sum += nums[i+j];
```
nums[i+j]は少しパズルに感じ、for (int j = i; j < nums.size(); ++j)としてsum += nums[j]にした方が素直かなと思いました。

maeken4 replied:
同じく、ループの範囲・条件と添え字の関係をいったん考えないといけない分少し読みにくく感じました。
i + j < nums.size()などすると少し取りやすいかもしれません。

5ky7 committed:
Update step 1

Replaced 'Hash Map' with '累積和 with look up' for accuracy

5ky7 committed:
Fix time complexity in naive implementation description

Corrected the worst-case time complexity from O(n^2) to O(n^3) for the naive implementation of the subsequence sum problem.

maeken4 commented on memo.md:
```diff
@@ -1,3 +1,148 @@
+        - o4-mini-highに聞いた結果：
+          - `sum`を用いるとjループを回すにはj-1のループが必要になるが，`sum`を用いない後者の方はjのループにj-1のループの結果が必要ないので並列化が可能．（データ依存の有無，コンパイラ最適化の受けやすさ）
+          - 3つ目のコードについて，`cumulative_sums[end]`のアクセスにおいてポインタインクリメントでアドレスが求まるので速いらしい
+            - だったら2つ目のコードもjの取り方を変えて`nums[i+j]`を`nums[j]`にしたら速くなる？ --> **速くなった！** Acceptされた（2700msくらい）．
```
時間計算量は同じでもデータ依存性で大きく実行時間が変わる例を見れて勉強になりました！
気になってこちらの手元でも計測してみたのですが、たしかにcode2とcode3で大きく変わりますね。ただ、code2の添え字をずらす分に関しては比較的小さいようでした。
各ループで添え字を計算する分違いは出ると思うのですが、データ依存性のほうが影響が大きいのかなと思いました。

参考までに以下結果と検証コードです。
```
 g++ -O3 -march=native -std=c++20 -DNDEBUG bench_subarray_sum.cpp -o bench

N=1000, trials=10000, max_abs=500, k=50
subarraySum_code2(平均): 0.369352 ms
subarraySum_code2_fix(平均): 0.367032 ms
subarraySum_code3(平均): 0.0600748 ms
```
```cpp
// bench_subarray_sum_avg.cpp
#include <algorithm>
#include <chrono>
#include <cstdint>
#include <functional>
#include <iostream>
#include <random>
#include <string>
#include <vector>

int subarraySum_code2(std::vector<int>& nums, int k) {
    int count = 0;
    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int sum = 0;
        for (int j = 0; j < static_cast<int>(nums.size()) - i; ++j) {
            sum += nums[i + j];
            if (sum == k) ++count;
        }
    }
    return count;
}

int subarraySum_code2_fix(std::vector<int>& nums, int k) {
    int count = 0;
    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int sum = 0;
        for (int j = i; j < static_cast<int>(nums.size()); ++j) {
            sum += nums[j];
            if (sum == k) ++count;
        }
    }
    return count;
}

int subarraySum_code3(std::vector<int>& nums, int k) {
    int n = nums.size();
    std::vector<int> cumulative_sums(n + 1);
    cumulative_sums[0] = 0;
    for (int i = 1; i <= n; ++i) {
        cumulative_sums[i] = nums[i - 1] + cumulative_sums[i - 1];
    }

    int count = 0;
    for (int start = 0; start < n; ++start) {
        for (int end = start + 1; end <= n; ++end) {
            if (cumulative_sums[end] - cumulative_sums[start] == k) {
                ++count;
            }
        }
    }
    return count;
}

// 平均実行時間[ms]を返す
template <class F>
double bench_avg_ms(F&& f, int trials, std::vector<int>& data, int k, int* out_result) {
    double total_ms = 0.0;
    int last = 0;
    for (int t = 0; t < trials; ++t) {
        auto t0 = std::chrono::steady_clock::now();
        int result = f(data, k);
        auto t1 = std::chrono::steady_clock::now();
        last = result;
        total_ms += std::chrono::duration<double, std::milli>(t1 - t0).count();
    }
    if (out_result) *out_result = last;
    return total_ms / trials;
}

int main(int argc, char** argv) {
    int N = 3000;
    int trials = 50;
    int max_abs = 10;
    int k = 0;

    if (argc >= 2) N = std::stoi(argv[1]);
    if (argc >= 3) trials = std::stoi(argv[2]);
    if (argc >= 4) max_abs = std::stoi(argv[3]);
    if (argc >= 5) k = std::stoi(argv[4]);

    std::mt19937 rng(123456789u);
    std::uniform_int_distribution<int> dist(-max_abs, max_abs);

    std::vector<int> a(N);
    for (int i = 0; i < N; ++i) a[i] = dist(rng);

    int r1 = 0, r2 = 0, r3 = 0;
    double t1 = bench_avg_ms(subarraySum_code2, trials, a, k, &r1);
    double t2 = bench_avg_ms(subarraySum_code2_fix, trials, a, k, &r2);
    double t3 = bench_avg_ms(subarraySum_code3, trials, a, k, &r3);

    const double subarrays = (static_cast<double>(N) * (N + 1)) / 2.0;
    std::cout << "N=" << N << ", trials=" << trials << ", max_abs=" << max_abs << ", k=" << k << "\n";
    std::cout << "subarraySum_code2(平均): " << t1 << " ms\n";
    std::cout << "subarraySum_code2_fix(平均): " << t2 << " ms\n";
    std::cout << "subarraySum_code3(平均): " << t3 << " ms\n";

    volatile int guard = r1 + r2 + r3;
    (void)guard;

    return 0;
}

```

oda commented on memo.md:
```diff
@@ -1,3 +1,148 @@
+## 前置インクリメントと後置インクリメントについて．
+  - 疑問の発端は「後置の方が（私にとっては）可読性が高いが前置の方が速いよな，どうしようか」
+  - 基本前置の方が速いと思っていたが，[そうとは限らない](https://cpp.aquariuscode.com/preincriment-vs-postincriment)らしい．
+    - 確かにgccでのintインクリメントは後置の方が速いというデータがある（clang, vc++は前置の方が速い）
```
んー、測定誤差のように思いますね。この記事を書いた人も 10 ms 程度は誤差と思っているように思います。
https://web.archive.org/web/20151226173448/http://resemblances.click3.org/?p=1828
https://web.archive.org/web/20160508102257/http://resemblances.click3.org/product_list/index.cgi/files/82/main.cpp

ボトルネックになるか云々は妥当かと思います。
